src/types.ts

```diff
@@ -0,0 +1,221 @@
+   *
+   * This namespace is used to consume the configuration from `window` and `localstorage`
+   */
+  namespace: string;
+
+  /**
+   * Schema of the configuration (used for runtime validation)
+   */
+  schema: TSchema;
+
+  /**
+   * Storage adapter
+   *
+   * @default window.localStorage
+   */
+  storage?: Storage;
+
+  /**
+   * Permit to overidde any config values in storage
+   *
+   * @default true
+   */
+  localOverride?: boolean;
+
+  /**
+   * Namespace for `useConfig()` return methods.
+   *
+   * Example:
+   * ```
+   *  // MyConfig.ts
+   *  export const { useConfig } = createConfig({
+   *    useConfigNamespace: "hello"
+   *   });
+   *
+   * // In a react component
+   * const {
+   *   getHelloConfig,
+   *   setHelloConfig,
+   *   getAllHelloConfig,
+   * } = useConfig();
+   * ```
+   */
+  useConfigNamespace?: TNamespace;
+}
+
+export type Config = StringConfig | NumberConfig | BooleanConfig | CustomConfig;
+
+export interface StringConfig {
+  type: "string";
+  enum?: string[];
+  default?: string | (() => string);
+  description?: string;
+}
+
+export interface StringEnumConfig extends StringConfig {
+  /**
+   * List of allowed values
+   */
+  enum: string[];
+}
+
+export interface NumberConfig {
+  type: "number";
+  min?: number;
+  max?: number;
+  default?: number | (() => number);
+  description?: string;
+}
+
+export interface BooleanConfig {
+  type: "boolean";
+  default?: boolean | (() => boolean);
+  description?: string;
+}
+
+export interface CustomConfig<T = unknown> {
+  type: "custom";
+  default?: T | (() => T);
+  description?: string;
+  /**
+   * Custom parser.
+   *
+   * Should throw an error if the value can't be parsed
+   */
+  parser: (value: any) => T;
+}
+
+export type ResolvedSchema<TSchema extends Record<string, Config>> = {
+  [key in keyof TSchema]: ResolvedConfigValue<TSchema[key]>;
+};
+
+export type ResolvedConfigValue<TValue extends Config> = TValue extends StringEnumConfig
+  ? TValue["enum"][-1]
+  : TValue extends StringConfig
+  ? string
+  : TValue extends NumberConfig
+  ? number
+  : TValue extends BooleanConfig
+  ? boolean
+  : TValue extends CustomConfig
+  ? ReturnType<TValue["parser"]>
+  : never;
+
+export const isStringConfig = (config: Config): config is StringConfig => config.type === "string";
+export const isStringEnumConfig = (config: Config): config is StringEnumConfig =>
+  config.type === "string" && Array.isArray(config.enum);
+export const isNumberConfig = (config: Config): config is NumberConfig => config.type === "number";
+export const isBooleanConfig = (config: Config): config is BooleanConfig => config.type === "boolean";
+export const isCustomConfig = (config: Config): config is CustomConfig => config.type === "custom";
+
+export interface InjectedProps<
+  TSchema extends Record<string, Config>,
+  TNamespace extends string,
+  TConfig = ResolvedSchema<TSchema>
+> {
+  namespace: string;
+  useConfigNamespace: TNamespace;
```

Comment:
`use` prefix make it easy to confuse with a hook. Would naming it 
```suggestion
configNamespace: TNamespace;
```
convey enough needed context?